x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/metric/SecurityCacheMetrics.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.security.metric;
+
+import org.elasticsearch.common.cache.Cache;
+import org.elasticsearch.telemetry.metric.LongWithAttributes;
+import org.elasticsearch.telemetry.metric.MeterRegistry;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public final class SecurityCacheMetrics {
+
+    public static <K, V> List<AutoCloseable> registerAsyncCacheMetrics(MeterRegistry registry, Cache<K, V> cache, CacheType type) {
+        final List<AutoCloseable> metrics = new ArrayList<>();
+        metrics.add(
+            registry.registerLongAsyncCounter(
+                type.metricsPrefix + ".hit.total",
+                "Total number of cache hits.",
+                "count",
+                () -> new LongWithAttributes(cache.stats().getHits())
+            )
+        );
+        metrics.add(
+            registry.registerLongAsyncCounter(
+                type.metricsPrefix + ".miss.total",
+                "Total number of cache misses.",
+                "count",
+                () -> new LongWithAttributes(cache.stats().getMisses())
+            )
+        );
+        metrics.add(
+            registry.registerLongAsyncCounter(
+                type.metricsPrefix + ".eviction.total",
+                "Total number of cache evictions.",
+                "count",
+                () -> new LongWithAttributes(cache.stats().getEvictions())
+            )
+        );
+        metrics.add(
+            registry.registerLongGauge(
+                type.metricsPrefix + ".count.total",
+                "The current number of cache entries.",
```

Comment:
This is not for this PR, but have we thought about reporting on the `size` (in `bytes`) for these caches? I would think that's useful, but looking at the code, seems tricky to implement, since I believe that's done by making the cache weight correspond to the byte size of entries, and we don't have that for the API key hashes, and looks like a big change.

Reply:
++ That would be really nice. I don't see an easy way to support this right now, but definitely a good future improvement.